### PR TITLE
PServer recovery from checkpoint

### LINKED ISSUE #2671: Pserver recover from saved state.

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #2671 

Related: Pserver save state https://github.com/PaddlePaddle/Paddle/pull/2716

### CODE REVIEW

File: go/pserver/etcd_client.go [hunk 1]
```diff
@@ -186,3 +188,20 @@ func (e *EtcdClient) registerPserverEtcd(ctx context.Context) (int, error) {
 
 	return idx, nil
 }
+
+// GetCheckpoints gets the checkpoint information by the specified pserver id
+func (e *EtcdClient) GetCheckpointInfo(string idx) (string, error) {
```
Comment (helinwang) [inline go/pserver/etcd_client.go]: Maybe rename this function to `GetKey`, caller can compute and pass in the key.

File: go/pserver/service.go [hunk 2]
```diff
@@ -49,13 +53,58 @@ type Service struct {
 	optMap map[string]*optimizer
 }
 
+// Checkpoint saves the checkpoint for pserver
+type Checkpoint struct {
+	UUID      string `json:"uuid"`
+	MD5       string `json:"md5"`
+	Timestamp string `json:"timestamp"`
+	State     []byte
```
Comment (helinwang) [inline go/pserver/service.go]: Just curious why State and ParameterWithConfig does not have json tag?

File: go/pserver/service.go [hunk 3]
```diff
@@ -49,13 +53,58 @@ type Service struct {
...
+	ParameterWithConfig
```
Comment (helinwang) [inline go/pserver/service.go]: Agree with your comment on the other PR, `Checkpoint` need to contain checkpoints for all parameters on the pserver.
I was confused before. I think `Checkpoint` should be renamed to `CheckpointInfo`, it's actually a reference to the checkpoint file on distributed filesystem. The type of `Timestamp` should be changed to `int64`.
Perhaps something like:
```go
type CheckpointInfo struct {
  UUID string
  Path string
  MD5 string
  Timestamp int64
}
```

Maybe we can have another type like:
```go
type ParameterCheckpoint struct {
  ParameterWithConfig
  State []byte
}

type Checkpoint []ParameterCheckpoint
```

How do you think?

Comment (Yancey1989) [inline go/pserver/service.go]: Good idea! I think separate `Checkpoint` and `CheckpointInfo` is a good idea.I will do that.

File: go/pserver/service.go (hunk 2 above)
Comment (Yancey1989) [inline go/pserver/service.go]: `Checkpoint` will be initialized from the JSON data in etcd, which only the the key `uuid`, `md5` and `timestamp`, but according with your other comment, separate `Checkpoint` and `CheckpointInfo` is better way, thanks for @helinwang .

File: go/pserver/service.go (hunk 3 above)
Comment (typhoonzero) [inline go/pserver/service.go]: Rename `ParameterCheckpoint` to `CheckpointData` should be more clear. Then these two struct should be:

```go
type CheckpointMeta struct {}
type CheckpointData struct {}
type Checkpoint struct {
  meta CheckpointMeta
  data CheckpointData
}
```

File: go/pserver/service.go [hunk 4]
```diff
@@ -49,13 +53,58 @@ type Service struct {
...
+}
+
+// NewCheckpoint creates a new checkpoint.
+func NewCheckpoint(idx int, cpPath string, e *EtcdClient) (*Checkpoint, error) {
```
Comment (typhoonzero) [inline go/pserver/service.go]: Rename to `NewCheckpointFromFile`

File: go/pserver/service.go [hunk 5]
```diff
@@ -49,13 +53,58 @@ type Service struct {
...
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	buf, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: create checkpoint from file
```
Comment (typhoonzero) [inline go/pserver/service.go]: Finish this TODO and then review.

File: go/pserver/service.go (hunk 3 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Cool, `CheckpointMeta` looks like a appropriate name.
Maybe we don't need the struct `Checkpoint` including `meta` and `data`? We can 
- `Json.Umarshal` from etcd to initialize `CheckpointMeta`.
- `gob.Decode(...)` from the checkpoint file to initialize `CheckpointData`

I looks like more simple?

File: go/pserver/service.go (hunk 4 above)
Comment (Yancey1989) [inline go/pserver/service.go]: As the offline discuss, I will add `Load` and `Save` method for the struct `Checkponit`:)

File: go/pserver/etcd_client.go (hunk 1 above)
Comment (Yancey1989) [inline go/pserver/etcd_client.go]: Done.

File: go/pserver/etcd_client.go [hunk 6]
```diff
@@ -186,3 +188,20 @@ func (e *EtcdClient) registerPserverEtcd(ctx context.Context) (int, error) {
...
+	key := PsCheckpoint + idx
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(timeout))
```
Comment (Yancey1989) [inline go/pserver/etcd_client.go]: Sorry :), fixed it.

File: go/pserver/service.go (hunk 2 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 3 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

@typhoonzero How do you think about this way?

File: go/pserver/service.go (hunk 4 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 3 above)
Comment (dzhwinter) [inline go/pserver/service.go]: agreed with @Yancey1989 . Do we need to separate `meta` and `data` in checkpoint ?

Comment (typhoonzero) [inline go/pserver/service.go]: Agree! 

> Do we need to separate meta and data in checkpoint ?

Maybe not, It's simple enouph to load meta and data into the struct separately.

File: go/pserver/service.go [hunk 7]
```diff
@@ -51,6 +56,88 @@ type Service struct {
...
+}
+
+// ParameterCheckpoint saves parameter checkpoint
+type ParameterCheckpoint struct {
+	ParameterWithConfig
+	Stat []byte
+}
+
+// Checkpoint saves all parameters' checkpoint
+type Checkpoint struct {
```
Comment (dzhwinter) [inline go/pserver/service.go]: maybe these things should not be the member of `Checkpoint`. They do not persist in permanent storage.  I move them into Server's argument.

File: go/pserver/service.go [hunk 8]
```diff
@@ -51,6 +56,88 @@ type Service struct {
...
+// CheckpointMeta saves the checkpoint information
+type CheckpointMeta struct {
+	UUID      string `json:"uuid"`
+	MD5       string `json:"md5"`
+	Timestamp string `json:"timestamp"`
+}
+
+// ParameterCheckpoint saves parameter checkpoint
+type ParameterCheckpoint struct {
+	ParameterWithConfig
```
Comment (Yancey1989) [inline go/pserver/service.go]: According with: https://golang.org/ref/spec#Struct_types, Field without a explicit field name is an anonymous field, so I can initialize the `Param` filed of `ParameterWithConfig` as `ParameterCheckpoint.Param = xxx`

File: go/pserver/service.go [hunk 9]
```diff
@@ -49,13 +54,95 @@ type Service struct {
 	optMap map[string]*optimizer
 }
 
+// CheckpointMeta saves the checkpoint information
+type CheckpointMeta struct {
```
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go [hunk 10]
```diff
@@ -49,13 +54,95 @@ type Service struct {
...
+	UUID      string `json:"uuid"`
+	MD5       string `json:"md5"`
+	Timestamp string `json:"timestamp"`
+}
+
+// ParameterCheckpoint saves parameter checkpoint
+type ParameterCheckpoint struct {
```
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 7 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go [hunk 11]
```diff
@@ -51,6 +56,88 @@ type Service struct {
...
+
+	f, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	dec := gob.NewDecoder(f)
+
+	if err = dec.Decode(&cp.data); err != nil {
```
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/cmd/pserver/pserver.go [hunk 12]
```diff
@@ -22,45 +23,52 @@ func main() {
...
 
-	s, err := pserver.NewService(idx)
-	if err != nil {
-		panic(err)
+	var s *pserver.Service
+	if newPserver {
+		s, err = pserver.NewService(idx)
+		candy.Must(err)
+	} else {
+		s, err = pserver.NewServiceFromCheckpoint(idx, cp)
```
Comment (Yancey1989) [inline go/cmd/pserver/pserver.go]: Done. Thx!

File: go/pserver/service.go (hunk 8 above)
Comment (dzhwinter) [inline go/pserver/service.go]: get it.  a brilliant feature of Go.

File: go/pserver/service.go [hunk 13]
```diff
@@ -51,21 +52,15 @@ type ParameterWithConfig struct {
...
 type checkpointMeta struct {
 	UUID      string `json:"uuid"`
-	Md5sum    string `json:"md5sum"`
+	MD5       string `json:"md5"`
 	Timestamp string `json:"timestamp"`
 }
 
 // Checkpoint is the pserver shard persist in file
-type Checkpoint []ParameterCheckpoint
+type Checkpoint []parameterCheckpoint
```
Comment (dzhwinter) [inline go/pserver/service.go]: I am following @typhoonzero 's comment here for possible usage. 
https://github.com/PaddlePaddle/Paddle/pull/2716#discussion_r126294140
But it is true, single parameterCheckpoint is useless to the user. It should be use immediately out of pserver's package.

File: go/pserver/service.go [hunk 14]
```diff
@@ -81,30 +76,78 @@ type Service struct {
...
+	}
+	defer f.Close()
+
+	h := md5.New()
+	if _, err = io.Copy(h, f); err != nil {
+		return nil, err
+	}
+	md5 := hex.EncodeToString(h.Sum(nil))
+	if md5 != cpMeta.MD5 {
+		return nil, errors.New("md5 does match, load checkpoint failed")
```
Comment (dzhwinter) [inline go/pserver/service.go]: do we need to put the const error type together? see the line of `L28`.

Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 13 above)
Comment (Yancey1989) [inline go/pserver/service.go]: We can fix this if something error for the later.

File: go/pserver/service.go [hunk 15]
```diff
@@ -26,6 +27,8 @@ const (
 	AlreadyInitialized = "pserver already initialized"
 	// Uninitialized is true if pserver not fully initialized
 	Uninitialized = "pserver not fully initialized"
+	// CheckpointUnmatched is true if the md5 of checkpoint file does matched the older
```
Comment (helinwang) [inline go/pserver/service.go]: I think "is true" is not very appropriate in English grammar, it's `string` type, "is true" is for `bool` type. Since the content of these constants already express the meaning, maybe we don't need the comment?
Perhaps we can put a comment on the top of `const` keyword:
```Go
// RPC error messages.
const (
  AlreadyInitialized = "pserver already initialized"
  Uninitialized = "pserver not fully initialized"
  CheckpointUnmatched = "checkpoint file does not matched the older"
)
```

File: go/pserver/service.go [hunk 16]
```diff
@@ -51,21 +54,15 @@ type ParameterWithConfig struct {
...
-	State       []byte
-}
-
-// checkpoint signature
+// checkpointMeta saves checkpoint metadata
 type checkpointMeta struct {
 	UUID      string `json:"uuid"`
-	Md5sum    string `json:"md5sum"`
+	MD5       string `json:"md5"`
 	Timestamp string `json:"timestamp"`
```
Comment (helinwang) [inline go/pserver/service.go]: Sorry this is not related to this PR.
`Timestamp` should have type `int64`, it's value should be from `func (t Time) UnixNano() int64`.
Having `string` here makes it not precise enough (精度不够).
Can you add a `TODO: change type of timestamp to int64` here? (or would be great if you could fix it 👍 ).

File: go/pserver/service.go [hunk 17]
```diff
@@ -81,30 +78,78 @@ type Service struct {
...
+	h := md5.New()
+	if _, err = io.Copy(h, f); err != nil {
+		return nil, err
+	}
+	md5 := hex.EncodeToString(h.Sum(nil))
+	if md5 != cpMeta.MD5 {
+		return nil, errors.New(CheckpointUnmatched)
+	}
+
+	dec := gob.NewDecoder(f)
```
Comment (helinwang) [inline go/pserver/service.go]: Since `io.Copy` already read through `f` once, `gob.NewDecoder(f)` requires to read from the beginning of the file, but `f` have not been `seek` to the beginning of the file. Does this actually work?

File: go/pserver/service.go [hunk 18]
```diff
@@ -81,30 +78,78 @@ type Service struct {
...
+		return nil, err
+	}
+	return cp, nil
+}
+
 // NewService creates a new service, will bypass etcd registration if no
-// endpoints specified.
-func NewService(idx int, seconds int, path string, client *EtcdClient, cp Checkpoint) (*Service, error) {
+// endpoints specified. It will recovery from checkpoint file if a exists a specified checkpoint.
+func NewService(idx int, interval int, path string, client *EtcdClient, cp *Checkpoint) (*Service, error) {
```
Comment (helinwang) [inline go/pserver/service.go]: Sorry it's not related to this PR. I know etcd uses `int` to express the type of timeout, we should really use type `time.Duration` for `interval`.

File: go/pserver/service.go [hunk 19]
```diff
@@ -186,13 +231,13 @@ func (s *Service) doCheckpoint() error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	cp := make([]ParameterCheckpoint, 0, len(s.optMap))
+	cp := make([]parameterCheckpoint, 0, len(s.optMap))
```
Comment (helinwang) [inline go/pserver/service.go]: This does not work. It creates a slice with length set to 0 and capacity set to `len(s.optMap)`. access `cp[0]` will crash the program since it's length is 0. Should use `cp := make([]parameterCheckpoint, len(s.optMap))` here.

For example:

```go
package main

func main() {
	a := make([]int, 0, 10)
	for i := 0; i < 10; i++ {
		a[i] = i
	}
}
```
Output:
```text
panic: runtime error: index out of range

goroutine 1 [running]:
main.main()
	/tmp/sandbox064034874/main.go:6 +0x20
```

File: go/pserver/etcd_client.go [hunk 20]
```diff
@@ -189,6 +189,22 @@ func (e *EtcdClient) registerPserverEtcd(ctx context.Context) (int, error) {
 	return idx, nil
 }
 
+// GetKey gets the value by the specified key
+func (e *EtcdClient) GetKey(key string, timeout int) ([]byte, error) {
```
Comment (Yancey1989) [inline go/pserver/etcd_client.go]: Done.

File: go/pserver/service.go (hunk 15 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go [hunk 21]
```diff
@@ -26,6 +27,8 @@ const (
 	AlreadyInitialized = "pserver already initialized"
 	// Uninitialized is true if pserver not fully initialized
 	Uninitialized = "pserver not fully initialized"
+	// CheckpointUnmatched is true if the md5 of checkpoint file does matched the older
+	CheckpointUnmatched = "checkpoint file does not matched the older"
```
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 16 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 17 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Sorry, it's a bug...  
Done.

File: go/pserver/service.go (hunk 18 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/service.go (hunk 19 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Thanks @helinwang , done.

File: go/cmd/pserver/pserver.go [hunk 22]
```diff
@@ -27,44 +28,39 @@ func main() {
...
 	} else {
 		timeout := time.Second * time.Duration((*etcdTimeout))
 		e = pserver.NewEtcdClient(*etcdEndpoint, *numPservers, timeout)
 		idx, err = e.Register()
+		candy.Must(err)
+
+		cp, err = pserver.NewCheckpointFromFile(*checkpointPath, idx, e)
 		if err != nil {
-			panic(err)
+			log.Infof("Fetch checkpoint failed, %s\n", err)
```
Comment (helinwang) [inline go/cmd/pserver/pserver.go]: Maybe `log.Errorf`?

File: go/cmd/pserver/pserver.go [hunk 23]
```diff
@@ -27,44 +28,39 @@ func main() {
...
 		}
 	}
 
-	s, err := pserver.NewService(idx, *checkpointInterval, *checkpointPath, e, cp)
-	if err != nil {
-		panic(err)
-	}
+	s, err := pserver.NewService(idx, time.Duration(*checkpointInterval)*time.Second, *checkpointPath, e, cp)
```
Comment (helinwang) [inline go/cmd/pserver/pserver.go]: Can change `flag.Int("checkpoint-interval", 600, "save checkpoint per interval seconds")` to `flag.Duration("checkpoint-interval", 600*time.Second, "save checkpoint interval")`, and not do type cast here.

File: go/pserver/service.go [hunk 24]
```diff
@@ -206,20 +244,24 @@ func (s *Service) doCheckpoint() error {
...
 	cpMeta.UUID = s.checkpointPath + strconv.Itoa(s.idx)
-	cpMeta.Timestamp = time.Now().String()
+	cpMeta.Timestamp = time.Now().UnixNano()
 	h := md5.New()
-	cpMeta.Md5sum = hex.EncodeToString(h.Sum(buf.Bytes()))
+	cpMeta.MD5 = hex.EncodeToString(h.Sum(buf.Bytes()))
 
 	cpMetajson, _ := json.Marshal(cpMeta)
-	err = s.client.PutKey(filepath.Join(PsCheckpoint, strconv.Itoa(s.idx)), cpMetajson, 3)
+	err = s.client.PutKey(filepath.Join(PsCheckpoint, strconv.Itoa(s.idx)), cpMetajson, time.Duration(3)*time.Second)
```
Comment (helinwang) [inline go/pserver/service.go]: Maybe change to `3*time.Second`. Constant in Go is well designed: https://blog.golang.org/constants

File: go/pserver/service.go [hunk 25]
```diff
@@ -81,30 +76,73 @@ type Service struct {
...
+			}
+			s.optMap[p.Param.Name] = newOptimizer(p, item.State)
 		}
 	}
 	return s, nil
 }
 
 // InitParam initializes a parameter.
-func (s *Service) InitParam(paramWithConfigs ParameterWithConfig, dummy *int) error {
+func (s *Service) InitParam(paramWithConfigs *ParameterWithConfig, dummy *int) error {
```
Comment (helinwang) [inline go/pserver/service.go]: Is passing pointer `* ParameterWithConfig` necessary? Please see: https://github.com/golang/go/wiki/CodeReviewComments#pass-values

File: go/pserver/optimizer.go [hunk 26]
```diff
@@ -35,7 +35,7 @@ func cArrayToSlice(p unsafe.Pointer, len int) []byte {
 	return (*[1 << 30]byte)(p)[:len:len]
 }
 
-func newOptimizer(paramWithConfigs ParameterWithConfig, State []byte) *optimizer {
+func newOptimizer(paramWithConfigs *ParameterWithConfig, State []byte) *optimizer {
```
Comment (helinwang) [inline go/pserver/optimizer.go]: Is passing pointer `* ParameterWithConfig` necessary? Please see: https://github.com/golang/go/wiki/CodeReviewComments#pass-values

File: go/cmd/pserver/pserver.go (hunk 22 above)
Comment (Yancey1989) [inline go/cmd/pserver/pserver.go]: Done.

File: go/cmd/pserver/pserver.go (hunk 23 above)
Comment (Yancey1989) [inline go/cmd/pserver/pserver.go]: Done.

File: go/pserver/service.go (hunk 25 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Cool, and here is also some useful documents about *pass by value of pointer*,
- http://goinbigdata.com/golang-pass-by-pointer-vs-pass-by-value/
- http://blog.rocana.com/golang-escape-analysis

File: go/pserver/service.go (hunk 24 above)
Comment (Yancey1989) [inline go/pserver/service.go]: Done.

File: go/pserver/optimizer.go (hunk 26 above)
Comment (Yancey1989) [inline go/pserver/optimizer.go]: Done.
